x/projects/keeper/creation.go

```diff
@@ -25,35 +27,91 @@ func (k Keeper) CreateProject(ctx sdk.Context, subscriptionAddress string, proje
 		return utils.LavaError(ctx, ctx.Logger(), "CreateEmptyProject_already_exist", map[string]string{"subscription": subscriptionAddress}, "project already exist for the current subscription with the same name")
 	}
 
-	project.Policy.EpochCuLimit = cuPerEpoch
-	project.Policy.TotalCuLimit = totalCU
-	project.Policy.MaxProvidersToPair = providers
+	project.Policy.EpochCuLimit = plan.GetComputeUnitsPerEpoch()
+	project.Policy.TotalCuLimit = plan.GetComputeUnits()
+	project.Policy.MaxProvidersToPair = plan.GetMaxProvidersToPair()
 	project.Policy.GeolocationProfile = geolocation
 
-	project.AppendKey(types.ProjectKey{Key: adminAddress, Types: []types.ProjectKey_KEY_TYPE{types.ProjectKey_ADMIN}, Vrfpk: vrfpk})
-
-	err := k.RegisterDeveloperKey(ctx, adminAddress, project.Index, blockHeight, vrfpk)
+	project.Enabled = enabled
+	project.Description = projectDescription
+	err := k.RegisterKey(ctx, types.ProjectKey{Key: adminAddress, Types: []types.ProjectKey_KEY_TYPE{types.ProjectKey_ADMIN, types.ProjectKey_DEVELOPER}, Vrfpk: vrfpk}, &project, blockHeight)
 	if err != nil {
 		return err
 	}
-
-	project.Enabled = enable
 	return k.projectsFS.AppendEntry(ctx, project.Index, blockHeight, &project)
 }
 
-func (k Keeper) RegisterDeveloperKey(ctx sdk.Context, developerKey string, projectIndex string, blockHeight uint64, vrfpk string) error {
+func (k Keeper) RegisterKey(ctx sdk.Context, key types.ProjectKey, project *types.Project, blockHeight uint64) error {
+	if project == nil {
+		return utils.LavaError(ctx, k.Logger(ctx), "RegisterKey_project_is_nil", nil, "project is nil")
+	}
+
+	developerKeyFound := false
+	inputKeyIsAdmin := false
+	inputKeyIsDeveloper := false
 	var developerData types.ProtoDeveloperData
-	_, found := k.developerKeysFS.FindEntry(ctx, developerKey, blockHeight, &developerData)
-	// a developer key with this address is not registered, add it to the developer keys list
-	if !found {
-		developerData.ProjectID = projectIndex
-		developerData.Vrfpk = vrfpk
-		err := k.developerKeysFS.AppendEntry(ctx, developerKey, blockHeight, &developerData)
-		if err != nil {
-			return err
+
+	for _, keyType := range key.GetTypes() {
+		if keyType == types.ProjectKey_ADMIN {
+			inputKeyIsAdmin = true
+		}
+
+		if keyType == types.ProjectKey_DEVELOPER {
+			inputKeyIsDeveloper = true
+			_, found := k.developerKeysFS.FindEntry(ctx, key.GetKey(), blockHeight, &developerData)
+			if found {
+				developerKeyFound = true
+			}
 		}
 	}
 
+	// handle admin key type
+	if inputKeyIsAdmin {
```

Comment:
do you depend on the iterations to be first developerkeytype and than adminkeytype? this is not a good way to handle this

Reply:
agreed. I think the logic should be:

```
existingKey = find inputKey.address
if not found {
    append inputKey to project
    done
}

// note: by now we know that the key is registered -
// if it is registered in another project return error
if existingKey.project != project {
    return error
}

// note: by now we know the key is already registered in this project
// also, inputKey has only one type in the bitmap (either admin or developer)
// so if existingKey already has this type then it already exists: do nothing
if existingKey.HasType(inputKey.type) {
    done
}

// note: by now we know that 
// -EITHER- inputKey is admin and existingKey is only developer, 
// -OR- inputKey is developer and existingKey is only admin
// so it's safe to add inputKey as is (will not override)
existingKey.type |= inputKey.type
append existingKey to project

done
```

Reply:
done